File: lib/data/models/home_status.dart
/// What is already running for the user, shown as the status pills above the deck.
///
/// Everything here is tolerant of missing fields: the pills are a glance, and a
/// backend that has not shipped a section yet should cost one pill, never the row.
class HomeStatus {
  final bool briefingReady;
  final int remindersToday;
  final List<HomeStatusRitual> rituals;
  final List<HomeStatusTracker> trackers;

  /// How many rituals are active, and how many the server will keep. Separate
  /// from [rituals], which is capped at three for the pill row and so can never
  /// prove the user is at the limit.
  ///
  /// Both default to 0, which reads as "unknown" rather than "at the cap": an
  /// older backend does not send these, and that must not lock the ritual card.
  final int ritualsActive;
  final int ritualsMax;

  const HomeStatus({
    this.briefingReady = false,
    this.remindersToday = 0,
    this.rituals = const [],
    this.trackers = const [],
    this.ritualsActive = 0,
    this.ritualsMax = 0,
  });

  static const HomeStatus empty = HomeStatus();

  bool get isEmpty =>
      !briefingReady &&
      remindersToday == 0 &&
      rituals.isEmpty &&
      trackers.isEmpty;

  factory HomeStatus.fromJson(Map<String, dynamic> json) {
    final rawRituals = json['rituals'];
    final rawTrackers = json['trackers'];
    return HomeStatus(
      briefingReady: json['briefing_ready'] == true,
      remindersToday:
          json['reminders_today'] is int ? json['reminders_today'] as int : 0,
      rituals: rawRituals is List
          ? rawRituals
              .map(HomeStatusRitual.fromJson)
              .whereType<HomeStatusRitual>()
              .toList()
          : const [],
      trackers: rawTrackers is List
          ? rawTrackers
              .map(HomeStatusTracker.fromJson)
              .whereType<HomeStatusTracker>()
              .toList()
          : const [],
      ritualsActive:
          json['rituals_active'] is int ? json['rituals_active'] as int : 0,
      ritualsMax: json['rituals_max'] is int ? json['rituals_max'] as int : 0,
    );
  }

  Map<String, dynamic> toJson() => {
        'briefing_ready': briefingReady,
        'reminders_today': remindersToday,
        'rituals': rituals.map((r) => r.toJson()).toList(),
        'trackers': trackers.map((t) => t.toJson()).toList(),
        'rituals_active': ritualsActive,
        'rituals_max': ritualsMax,
      };
}

class HomeStatusRitual {
  final String ritualId;
  final String title;

  /// Local wall time of the next occurrence, as the server resolved it.
  final String nextFireLocal;

  const HomeStatusRitual({
    required this.ritualId,
    required this.title,
    this.nextFireLocal = '',
  });

  /// The short form for a pill: "Joke at 8:00 AM".
  String get pillLabel {
    final clock = _clockOf(nextFireLocal);
    final subject = title.split(' ').take(2).join(' ');
    return clock.isEmpty ? title : '$subject at $clock';
  }

  static String _clockOf(String isoLocal) {
    if (isoLocal.isEmpty) return '';
    final parsed = DateTime.tryParse(isoLocal);
    if (parsed == null) return '';
    final hour = parsed.hour % 12 == 0 ? 12 : parsed.hour % 12;
    final minute = parsed.minute.toString().padLeft(2, '0');
    return '$hour:$minute ${parsed.hour < 12 ? 'AM' : 'PM'}';
  }

  static HomeStatusRitual? fromJson(Object? raw) {
    if (raw is! Map) return null;
    final id = raw['ritual_id'];
    final title = raw['title'];
    if (id is! String || id.isEmpty || title is! String || title.isEmpty) {
      return null;
    }
    return HomeStatusRitual(
      ritualId: id,
      title: title,
      nextFireLocal:
          raw['next_fire_local'] is String ? raw['next_fire_local'] as String : '',
    );
  }

  Map<String, dynamic> toJson() => {
        'ritual_id': ritualId,
        'title': title,
        'next_fire_local': nextFireLocal,
      };
}

class HomeStatusTracker {
  final String trackerId;
  final String title;

  const HomeStatusTracker({required this.trackerId, required this.title});

  static HomeStatusTracker? fromJson(Object? raw) {
    if (raw is! Map) return null;
    final id = raw['tracker_id'];
    final title = raw['title'];
    if (id is! String || id.isEmpty || title is! String || title.isEmpty) {
      return null;
    }
    return HomeStatusTracker(trackerId: id, title: title);
  }

  Map<String, dynamic> toJson() => {
        'tracker_id': trackerId,
        'title': title,
      };
}

File: backend/src/handlers/home_status.py
"""GET /home/status — everything the home screen's status pills need, in one call.

Four separate calls on app open would be four round-trips for one strip of text,
so this gathers them in parallel and answers once. Nothing here is allowed to be
expensive: no model call, no full briefing payload, and no query that needs an
index the project does not already have.

Every section is isolated. A failure in one of them yields that field empty
rather than a 500, because a missing pill is a smaller problem than a home screen
that cannot finish loading.
"""

from __future__ import annotations

import asyncio
from datetime import UTC, datetime
from typing import Any
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from fastapi import Request
from fastapi.responses import JSONResponse

from ..lib.logger import logger
from ..services import rituals
from ..services.briefing import fields as bf
from ..services.firebase import admin_firestore
from ..services.request_auth import resolve_user_id_from_request
from ..services.tracking import fields as tf
from ..services.tracking import tracking_store

# Enough to cover any realistic day for one user without an unbounded scan. The
# query is single-field (status ==), which Firestore auto-indexes, so this stays
# cheap; the day filter happens in memory for the same reason the alarm-sync path
# does it that way (handlers/reminders.py) — a composite index on a path the home
# screen depends on is one deployment away from silently returning nothing.
_PENDING_REMINDER_SCAN_LIMIT = 200

# Pills are a glance, not a list.
_MAX_RITUAL_PILLS = 3
_MAX_TRACKER_PILLS = 3


def _user_timezone(user_id: str) -> str:
    """One read of users/{uid}, reused for both the briefing date and the day window."""
    try:
        snap = admin_firestore().collection("users").document(user_id).get()
        if snap.exists:
            value = (snap.to_dict() or {}).get("timezone")
            if isinstance(value, str) and value.strip():
                return value.strip()
    except Exception as exc:
        logger.warn("home_status: timezone read failed, defaulting UTC", {
            "user_id": user_id, "error": str(exc),
        })
    return "UTC"


def _local_now(timezone_name: str) -> datetime:
    try:
        return datetime.now(ZoneInfo(timezone_name))
    except (ZoneInfoNotFoundError, ValueError):
        return datetime.now(UTC)


def _briefing_ready(user_id: str, local_date: str) -> bool:
    """Is today's briefing already generated?

    Deliberately NOT the /briefing/today path: that returns the whole narrative
    and every item, and the client's briefing viewmodel generates one on a miss.
    This reads a single document and only its status field.
    """
    try:
        snap = (
            admin_firestore()
            .collection("users")
            .document(user_id)
            .collection(bf.DAILY_BRIEFING_SUBCOLLECTION)
            .document(local_date)
            .get(field_paths=[bf.FIELD_STATUS])
        )
        if not snap.exists:
            return False
        return str((snap.to_dict() or {}).get(bf.FIELD_STATUS, "")) == bf.STATUS_READY
    except Exception as exc:
        logger.warn("home_status: briefing readiness read failed", {
            "user_id": user_id, "error": str(exc),
        })
        return False


def _reminders_today(user_id: str, local_now: datetime) -> int:
    """How many pending reminders land on the user's local calendar day."""
    try:
        docs = list(
            admin_firestore()
            .collection("users")
            .document(user_id)
            .collection("reminders")
            .where("status", "==", "pending")
            .limit(_PENDING_REMINDER_SCAN_LIMIT)
            .stream()
        )
    except Exception as exc:
        logger.warn("home_status: reminder scan failed", {
            "user_id": user_id, "error": str(exc),
        })
        return 0

    today = local_now.date()
    zone = local_now.tzinfo or UTC
    count = 0
    for doc in docs:
        raw = (doc.to_dict() or {}).get("trigger_at")
        if not isinstance(raw, str) or not raw.strip():
            continue
        try:
            when = datetime.fromisoformat(raw.strip().replace("Z", "+00:00"))
        except ValueError:
            continue
        when = when if when.tzinfo else when.replace(tzinfo=UTC)
        if when.astimezone(zone).date() == today:
            count += 1
    return count


async def _ritual_pills(user_id: str) -> dict[str, Any]:
    """The pills, plus the counts behind them.

    The pill list is capped at three, so it can never tell the client whether the
    user is at the ritual limit. The counts are returned alongside it precisely so
    the home deck can stop offering a ritual it knows the server would refuse.
    """
    items = await rituals.list_rituals(user_id)
    active = [item for item in items if str(item.get("status")) == rituals.STATUS_ACTIVE]
    active.sort(key=lambda item: str(item.get("next_fire_at") or ""))
    return {
        "pills": [
            {
                "ritual_id": str(item.get("id", "")),
                "title": str(item.get("title", "")),
                "next_fire_local": str(item.get("next_fire_local", "")),
            }
            for item in active[:_MAX_RITUAL_PILLS]
        ],
        "active": len(active),
        "max": rituals.MAX_ACTIVE_RITUALS,
    }


async def _tracker_pills(user_id: str) -> list[dict[str, Any]]:
    trackers = await tracking_store.list_trackers_for_user(user_id)
    active = [t for t in trackers if t.status == tf.TRACKER_STATUS_ACTIVE]
    pills: list[dict[str, Any]] = []
    for tracker in active[:_MAX_TRACKER_PILLS]:
        topic = await tracking_store.get_tracked_topic(tracker.topic_key)
        title = (topic.title if topic else "") or tracker.topic_key.replace("-", " ")
        pills.append({"tracker_id": tracker.id, "title": title})
    return pills


async def handle_home_status(request: Request) -> JSONResponse:
    user_id = resolve_user_id_from_request(request)
    if not user_id:
        return JSONResponse({"error": "Unauthorized"}, status_code=401)

    timezone_name = await asyncio.to_thread(_user_timezone, user_id)
    local_now = _local_now(timezone_name)

    briefing, reminders, ritual_state, tracker_pills = await asyncio.gather(
        asyncio.to_thread(_briefing_ready, user_id, local_now.date().isoformat()),
        asyncio.to_thread(_reminders_today, user_id, local_now),
        _ritual_pills(user_id),
        _tracker_pills(user_id),
        return_exceptions=True,
    )

    def _or_default(value: Any, default: Any, label: str) -> Any:
        if isinstance(value, BaseException):
            logger.warn("home_status: section failed", {
                "user_id": user_id, "section": label, "error": str(value),
            })
            return default
        return value

    ritual_state = _or_default(
        ritual_state,
        {"pills": [], "active": 0, "max": rituals.MAX_ACTIVE_RITUALS},
        "rituals",
    )

    return JSONResponse({
        "briefing_ready": bool(_or_default(briefing, False, "briefing")),
        "reminders_today": int(_or_default(reminders, 0, "reminders")),
        "rituals": ritual_state["pills"],
        "trackers": _or_default(tracker_pills, [], "trackers"),
        # A failed section reports 0 active, which the client reads as "unknown"
        # and keeps offering rituals. Letting a read failure look like being at
        # the cap would silently disable the card.
        "rituals_active": int(ritual_state["active"]),
        "rituals_max": int(ritual_state["max"]),
    })
